tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/jms/multiprotocol/JMSFQQNConsumerTest.java

```diff
@@ -0,0 +1,209 @@
+      server.getAddressSettingsRepository().clear();
+      server.getAddressSettingsRepository().addMatch("#", settings);
+      ConnectionFactory factory = CFUtil.createConnectionFactory(protocol, "tcp://localhost:5672");
+      final String queue = "queue";
+      final String address = "address";
+      final String filter = "prop='value'";
+      boolean thrownException = false;
+      try (Connection c = factory.createConnection()) {
+         Session s = c.createSession(false, Session.AUTO_ACKNOWLEDGE);
+         Topic t = s.createTopic(CompositeAddress.toFullyQualified(address, queue));
+         MessageConsumer mc = s.createConsumer(t);
+      } catch (Exception e) {
+         thrownException = true;
+      }
+
+      Assert.assertTrue(thrownException);
+   }
+
+   @Test
+   public void testFQQNQueueConsumerWithSelectorAMQP() throws Exception {
+      testFQQNQueueConsumerWithSelector("AMQP");
+   }
+
+   @Test
+   public void testFQQNQueueConsumerWithSelectorOpenWire() throws Exception {
+      testFQQNQueueConsumerWithSelector("OPENWIRE");
+   }
+
+   @Test
+   public void testFQQNQueueConsumerWithSelectorCore() throws Exception {
+      testFQQNQueueConsumerWithSelector("CORE");
+   }
+
+   private void testFQQNQueueConsumerWithSelector(String protocol) throws Exception {
+      AddressSettings settings = new AddressSettings().setDefaultQueueRoutingType(RoutingType.ANYCAST).setDefaultAddressRoutingType(RoutingType.ANYCAST);
+      server.getAddressSettingsRepository().addMatch("#", settings);
+      final String queue = "myQueue";
+      final String address = "address";
+      final String filter = "prop='inside'";
+      ConnectionFactory factory = CFUtil.createConnectionFactory(protocol, "tcp://localhost:5672");
+      try (Connection c = factory.createConnection()) {
+         c.start();
+         Session s = c.createSession(false, Session.AUTO_ACKNOWLEDGE);
+         String queueQuery = CompositeAddress.toFullyQualified(address, queue) + (protocol.equals("OPENWIRE") ? "?selectorAware=true" : "");
+         Queue q = s.createQueue(queueQuery);
+         MessageConsumer mc = s.createConsumer(q, filter);
+         Wait.assertTrue(() -> server.locateQueue(queue) != null, 2000, 100);
+         org.apache.activemq.artemis.core.server.Queue serverQueue = server.locateQueue(SimpleString.toSimpleString(queue));
+         Assert.assertEquals(RoutingType.ANYCAST, serverQueue.getRoutingType());
+         Assert.assertNull(serverQueue.getFilter());
+         MessageProducer p = s.createProducer(q);
+         Message m = s.createMessage();
+         m.setStringProperty("prop", "inside");
+         p.send(m);
+         assertNotNull(mc.receive(1000));
+         m = s.createMessage();
+         m.setStringProperty("prop", RandomUtil.randomString());
+         assertNull(mc.receiveNoWait());
+
+         serverQueue.getConsumers().forEach(queueConsumer -> {
+            Assert.assertNotNull(queueConsumer.getFilter());
```

Comment:
Can it match the actual filter?

To Roman's points, having more than 1 consumer with _different_ filters would seem useful in this test to verify they are handled correctly. Probably also sending a message that doesnt match any of them...check it remains, awaiting some later consumer on the queue.

Reply:
It should double check that the forEach was actually entered, e.g count the consumers....which again there remains only one, there should be at least 2 with different filters to verify they are handled correctly.

Reply:
this test will be only one.. there's another test I added that is validating filters.

It is validating the filters by matching the messages.

Reply:
> It should double check that the forEach was actually entered, e.g count the consumers....which again there remains only one, there should be at least 2 with different filters to verify they are handled correctly.

testFQQNTopicMultiConsumerWithSelector was added to validate multiple consumers.

Reply:
those are for topic usage...this one is queue usage

Reply:
and I'm not interested in validating the routing itself here.. I am closing that consumer, and opening a new one now.